File: 0155.MinStack.py
class MinStack(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.data = []
        self.min = []

    def push(self, x):
        """
        :type x: int
        :rtype: nothing
        """
        '''
        min_x = self.getMin()
        if min_x == None or x < min_x:
            min_x = x 
        self.data.append((x, min_x))
        '''
        self.data.append(x)
        if x <= self.getMin() or self.getMin() is None:
            self.min.append(x)

    def pop(self):
        """
        :rtype: nothing
        """
        '''
        if self.data:
            self.data.pop()
        '''
        if self.data[-1] == self.min[-1]:
            self.min.pop()
        self.data.pop()        
        
    def top(self):
        """
        :rtype: int
        """
        '''
        if self.data:
            return self.data[-1][0]
        else:
            return None
        '''
        if self.data:
            return self.data[-1]
            
    def getMin(self):
        """
        :rtype: int
        """
        '''
        if self.data:
            return self.data[-1][1]
        else:
            return None
        '''
        if self.min:
            return self.min[-1]
        else:
            return None

File: 0000-0999/0155.MinStack.go
type MinStack struct {
    stack []int
    min int
}


/** initialize your data structure here. */
func Constructor() MinStack {
    return MinStack{stack: []int{}, min: math.MaxInt32}
}


func (this *MinStack) Push(x int)  {
    if x <= this.min { // will WA if only "<"
        this.stack = append(this.stack, this.min)
        this.min = x
    }
    this.stack = append(this.stack, x)
}


func (this *MinStack) Pop()  {
    n := len(this.stack)
    if this.stack[n-1] == this.min {
        this.min = this.stack[n-2]
        this.stack = this.stack[:n-2]
    } else {
        this.stack = this.stack[:n-1]
    }
}


func (this *MinStack) Top() int {
    return this.stack[len(this.stack)-1]
}


func (this *MinStack) GetMin() int {
    return this.min
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */